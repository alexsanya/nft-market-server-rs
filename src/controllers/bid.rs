use tokio::join;
use crate::{models::bid::Bid, prelude::Result, repositories::bids::{get_all, save_bid}, services::onchain_provider::{ensure_bidder_has_tokens, ensure_erc20_allowance, ensure_owner_has_nft, esnure_erc712_allowance}};

pub async fn create_bid(bid: &Bid) -> Result<()> {
    bid.verify_signature()?;
    let (nft_ownership_check, nft_allowance_check, tokens_check, allowance_check)  = join!(
        ensure_owner_has_nft(&bid.listing.owner, &bid.listing.nft_contract, &bid.listing.token_id),
        esnure_erc712_allowance(&bid.listing.owner, &bid.listing.nft_contract),
        ensure_bidder_has_tokens(&bid.bidder, &bid.token_address, &bid.value),
        ensure_erc20_allowance(&bid.bidder, &bid.token_address, &bid.value)
    );
    nft_ownership_check?;
    nft_allowance_check?;
    tokens_check?;
    allowance_check?;
    save_bid(bid).await?;
    Ok(())
}

pub fn get_bids() -> Result<Vec<Bid>> {
    get_all()
}
