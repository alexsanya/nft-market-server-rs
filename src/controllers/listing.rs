use crate::{models::listing::Listing, repositories::listing::{get_all, save_listing}};


pub async fn create_listing(listing: &Listing) -> Result<(), ()> {
    save_listing(listing).await?;
    Ok(())
}

pub fn get_listings() -> Vec<Listing> {
    get_all().unwrap()
}