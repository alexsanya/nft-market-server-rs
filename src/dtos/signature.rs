
use serde::{Serialize, Deserialize};

use crate::{models::signature::Signature, utils::patterns::Patterns};

#[derive(Debug, Serialize, Deserialize)]
pub struct SignatureDTO {
    pub v: u64,
    pub r: String,
    pub s: String
}

#[derive(Debug, Clone)]
pub enum ParsingError {
    RisInvalid,
    SisInvalid
}

impl TryInto<Signature> for SignatureDTO {
    type Error = ParsingError;

    fn try_into(self) -> Result<Signature, Self::Error> {
        let patterns = Patterns::new();
        Ok(
            Signature{
                v: self.v,
                r: patterns.test_bytes32(&self.r).map_err(|_| ParsingError::RisInvalid)?.to_owned(),
                s: patterns.test_bytes32(&self.s).map_err(|_| ParsingError::SisInvalid)?.to_owned()
            }
        )
    }
}