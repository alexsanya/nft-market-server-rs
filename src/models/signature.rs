
use ethers::contract::Eip712;
use serde::{Serialize, Deserialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Signature {
    pub v: u64,
    pub r: String,
    pub s: String
}