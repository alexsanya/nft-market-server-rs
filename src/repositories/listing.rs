use serde_json;
use tracing::debug;
use crate::datasource::{set_value, get_all as fetch};
use crate::dtos::listing::ListingDTO;
use crate::models::listing::Listing;
use crate::error::Error;
use crate::prelude::Result as MyResult;

pub async fn save_listing(listing: &Listing) -> MyResult<()> {
    let key = format!("listings:{}", &listing.get_hash()?);
    let value = serde_json::to_string(listing).map_err(|_| Error::SaveData)?;
    debug!("Serrialized listing: {}", value);
    set_value(&key, &value).await?;
    Ok(())
}

pub async fn get_all() -> MyResult<Vec<Listing>> {
    let values = fetch("listings:*").await?;
    let listings = values.iter().filter_map(|value| {
        let listing_dto: Result<ListingDTO, _> = serde_json::from_str(&value);
        if let Ok(listing_dto) = listing_dto {
            match listing_dto.try_into() {
                Ok(listing) => Some(listing),
                _ => None
            }
        } else {
            None
        }
    }).collect();
    Ok(listings)
}