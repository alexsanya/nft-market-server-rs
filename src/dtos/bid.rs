use serde::{Deserialize, Serialize};
use crate::models::bid::Bid;
use crate::utils::patterns::Patterns;
use super::listing::ListingDTO;
use super::signature::SignatureDTO;
use super::listing::ParsingError as ParsingListingError;
use super::signature::ParsingError as ParsingSignatureError;

#[derive(Debug, strum_macros::AsRefStr)]
pub enum ParsingError {
    Bidder,
    TokenAddress,
    ValidUntil,
    Value,
    Listing(ParsingListingError),
    Signature(ParsingSignatureError)
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BidDTO {
    pub bidder: String,
    pub listing: ListingDTO,
    pub token_address: String,
    pub valid_until: String,
    pub value: String,
    pub signature: SignatureDTO
}

impl TryInto<Bid> for BidDTO {
    type Error = ParsingError;

    fn try_into(self) -> Result<Bid, Self::Error> {
        let patterns = Patterns::new();
        let bid = Bid {
            bidder: patterns.test_address(&self.bidder).map_err(|_| ParsingError::Bidder)?.to_owned(),
            listing: self.listing.try_into().map_err(ParsingError::Listing)?,
            token_address: patterns.test_address(&self.token_address).map_err(|_| ParsingError::TokenAddress)?.to_owned(),
            valid_until: self.valid_until.parse().map_err(|_| ParsingError::ValidUntil)?,
            value: self.value.parse().map_err(|_| ParsingError::Value)?,
            siganture: self.signature.try_into().map_err(ParsingError::Signature)?
        };
        Ok(bid)
    }
}

#[cfg(test)]
mod test {
    use serde_json::json;
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn convert() {
        let raw_listing = json!({
            "owner": "0x3897326cEda92B3da2c27a224D6fDCFefCaCf57A",
            "chain_id": "11155111",
            "min_price_cents": "150000",
            "nft_contract": "0xf44b599a0aB6b8cb14E992994BEC0dc59dF883B2",
            "token_id": "1",
            "nonce": "0",
            "signature": {
                "v": 28,
                "r": "0x5ef4620f4b296763ff15209456d75e868f149a8d1c6821f1ff11fab70bca0ee0",
                "s": "0x337ddcb26ea919a2bf5ad6e1d49bd6951a27d1d2e940a5543a70eabc5dbe237e"
            }
        });

        let raw_bid = json!({
            "bidder": "0xE98D94496aB9084f597a69978b593EBf83147335",
            "listing": raw_listing,
            "token_address": "0xc29f6F8D639eF187DcFEfeFBaD989cF2C941a23A",
            "valid_until": "1735504160",
            "value": "250",
            "signature": {
                "v": 28,
                "r": "0x1469ac6f9636c24d2d8c3fb2cbef73708876e15f23f23b1d33863939c905a21c",
                "s": "0x7d9a7ea039465c928311bcb737b23153232028038beadba2a667aa720f17602b"
            }
        });

        let bid_dto: BidDTO = serde_json::from_value(raw_bid).expect("Cannot convert jsonBid to BidDTO");
        let bid: Bid = bid_dto.try_into().expect("Cannot convert BidBTo to bid");
        assert_debug_snapshot!(bid);
    }
}