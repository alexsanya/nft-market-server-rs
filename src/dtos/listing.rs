use serde::{Serialize, Deserialize};
use crate::{models::listing::Listing, utils::patterns::Patterns};
use super::signature::{self, SignatureDTO};

#[derive(Debug)]
pub enum ParsingError {
    Owner,
    ChainId,
    NftContract,
    TokenId,
    Nonce,
    Signature(signature::ParsingError)
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ListingDTO {
    pub owner: String,
    pub chain_id: String,
    pub min_price_cents: u16,
    pub nft_contract: String,
    pub token_id: String,
    pub nonce: String,
    pub signature: SignatureDTO
}

impl TryInto<Listing> for ListingDTO {
    type Error = ParsingError;

    fn try_into(self) -> Result<Listing, Self::Error> {
        let patterns = Patterns::new();
        let listing = Listing{
            owner: patterns.test_address(&self.owner).map_err(|_| ParsingError::Owner)?.to_owned(),
            chain_id: self.chain_id.parse().map_err(|_| ParsingError::ChainId)?,
            min_price_cents: self.min_price_cents,
            nft_contract: patterns.test_address(&self.nft_contract).map_err(|_| ParsingError::NftContract)?.to_owned(),
            token_id: self.token_id.parse().map_err(|_| ParsingError::TokenId)?,
            nonce: self.nonce.parse().map_err(|_| ParsingError::Nonce)?,
            signature: self.signature.try_into().map_err(ParsingError::Signature)?
        };
        Ok(listing)
    }
}