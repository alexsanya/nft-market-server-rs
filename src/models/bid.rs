use num_bigint::BigInt;
use serde::{Deserialize, Serialize};
use super::signature::Signature as SigString;

use super::listing::Listing;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Bid {
    pub bidder: String,
    pub listing: Listing,
    pub token_address: String,
    pub valid_until: BigInt,
    pub value: BigInt,
    pub siganture: SigString
}