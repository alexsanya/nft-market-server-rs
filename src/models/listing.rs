use ethers::types::{Signature, U256};
use num_bigint::BigInt;
use ethers::abi::Address;
use std::str::FromStr;
use serde::{Serialize, Deserialize};
use crate::error::{Entity, Error};
use crate::prelude::Result as MyResult;

use super::listing_eip712::Listing as ListingEIP712;
use super::signature::Signature as SigString;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Listing {
    pub owner: String,
    pub chain_id: BigInt,
    pub min_price_cents: BigInt,
    pub nft_contract: String,
    pub token_id: BigInt,
    pub nonce: BigInt,
    pub signature: SigString
}

impl Listing {
    pub fn verify_signature(&self) -> MyResult<()> {
        let signature: Signature = self.signature.clone().try_into().map_err(|_| Error::InvalidSignature(Entity::Listing))?;
        let owner_result = Address::from_str(&self.owner);
        let listing_eip721: ListingEIP712 = self.clone().try_into().map_err(|_| Error::InvalidSignature(Entity::Listing))?;
        match (signature.recover_typed_data(&listing_eip721), owner_result) {
            (Ok(recovered), Ok(owner)) if recovered == owner => Ok(()),
            _ => Err(Error::InvalidSignature(Entity::Listing))
        }
    }
}

#[derive(Debug)]
pub enum Eip712Error {
    MinPrice,
    NftContract,
    TokenId,
    Nonce
}

impl TryInto<ListingEIP712> for Listing {
    type Error = Eip712Error;

    fn try_into(self) -> Result<ListingEIP712, Self::Error> {
        let listing_eip712 = ListingEIP712 {
            min_price_cents: U256::from_dec_str(&self.min_price_cents.to_string()).map_err(|_| Eip712Error::MinPrice)?,
            nft_contract: Address::from_str(&self.nft_contract).map_err(|_| Eip712Error::NftContract)?,
            token_id: U256::from_dec_str(&self.token_id.to_string()).map_err(|_| Eip712Error::TokenId)?,
            nonce: U256::from_dec_str(&self.nonce.to_string()).map_err(|_| Eip712Error::Nonce)?,
        };
        Ok(listing_eip712)
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;
    use ethers::types::Signature;
    use ethers::types::Address;
    use serde_json::json;
    use crate::dtos::listing::ListingDTO;
    use super::*;

    #[test]
    fn check_hash() {
        let raw_listing = json!({
            "owner": "0x3897326cEda92B3da2c27a224D6fDCFefCaCf57A",
            "chain_id": "11155111",
            "min_price_cents": "150000",
            "nft_contract": "0xf44b599a0aB6b8cb14E992994BEC0dc59dF883B2",
            "token_id": "1",
            "nonce": "0",
            "signature": {
                "v": 28,
                "r": "0x5ef4620f4b296763ff15209456d75e868f149a8d1c6821f1ff11fab70bca0ee0",
                "s": "0x337ddcb26ea919a2bf5ad6e1d49bd6951a27d1d2e940a5543a70eabc5dbe237e"
            }
        });

        let listing_dto: ListingDTO = serde_json::from_value(raw_listing).expect("Cannot convert json listing to ListingDTO");
        let listing: Listing = listing_dto.try_into().expect("Cannot convert ListingDTp to Listing");
        let signature: Signature = listing.signature.clone().try_into().expect("Cannot convert signature");
        println!("Signature: {}", signature);

        let listing_eip712: ListingEIP712 = listing.try_into().expect("Failed to convert listing into EIP712");
        println!("listing_eip721: {:?}", listing_eip712);
        //println!("EIP712 hash: {}", listing_eip712.encode_eip712().unwrap());

        let address = signature.recover_typed_data(&listing_eip712).expect("Cannot recover typed data");
        println!("Recovered address: {:?}", address);

        assert_eq!(address, Address::from_str("0x3897326ceda92b3da2c27a224d6fdcfefcacf57a").unwrap());

    }
}