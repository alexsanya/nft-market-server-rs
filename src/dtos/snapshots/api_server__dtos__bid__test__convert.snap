---
source: src/dtos/bid.rs
expression: bid
snapshot_kind: text
---
Bid {
    bidder: "0xE98D94496aB9084f597a69978b593EBf83147335",
    listing: Listing {
        owner: "0x3897326cEda92B3da2c27a224D6fDCFefCaCf57A",
        chain_id: 11155111,
        min_price_cents: 150000,
        nft_contract: "0xf44b599a0aB6b8cb14E992994BEC0dc59dF883B2",
        token_id: 1,
        nonce: 0,
        signature: Signature {
            v: 28,
            r: "0x5ef4620f4b296763ff15209456d75e868f149a8d1c6821f1ff11fab70bca0ee0",
            s: "0x337ddcb26ea919a2bf5ad6e1d49bd6951a27d1d2e940a5543a70eabc5dbe237e",
        },
    },
    token_address: "0xc29f6F8D639eF187DcFEfeFBaD989cF2C941a23A",
    valid_until: 1735504160,
    value: 250,
    siganture: Signature {
        v: 28,
        r: "0x1469ac6f9636c24d2d8c3fb2cbef73708876e15f23f23b1d33863939c905a21c",
        s: "0x7d9a7ea039465c928311bcb737b23153232028038beadba2a667aa720f17602b",
    },
}
